File: src/main/java/gargoyle/netsquares/model/Move.java
package gargoyle.netsquares.model;

public class Move {
    private Direction direction;
    private int distance;

    public Move(final Direction direction, final int distance) {
        super();
        setDirection(direction);
        setDistance(distance);
    }

    public Direction getDirection() {
        return direction;
    }

    public int getDistance() {
        return distance;
    }

    private void setDirection(final Direction direction) {
        this.direction = direction;
    }

    private void setDistance(final int distance) {
        this.distance = distance;
    }

    @Override
    public String toString() {
        return getClass().getSimpleName() + "[" + ("direction=" + getDirection()) + ","
                + ("distance=" + getDistance()) + "]";
    }
}

File: src/main/kotlin/gargoyle/netsquares/model/Move.kt
package gargoyle.netsquares.model

enum class Move(val direction: Direction?) {
    UP(Direction.UP), DOWN(Direction.DOWN), LEFT(Direction.LEFT), RIGHT(Direction.RIGHT), OPEN(null);

    fun isAllowed(allowed: Directions?): Boolean = this == OPEN || allowed!!.isAllowed(direction)

    override fun toString(): String = String.format("Move.${name}(${direction})")

    companion object {
        fun forDirection(direction: Direction?): Move? {
            for (move in values()) {
                if (move.direction == direction) {
                    return move
                }
            }
            return null
        }
    }
}
